File: javascript/algorithms/srtf.js
const createArrivalTimeComparator = require("../comparators/arrivalTime");
const createBurstTimeComparator = require("../comparators/burstTime");

const getProcessToExecute = (
  processes,
  currentExecutionTime,
  burstTimeComparator
) => {
  return processes
    .filter(
      process =>
        process.arrivalTime <= currentExecutionTime && process.burstTime !== 0
    )
    .sort(burstTimeComparator)[0];
};

const srtf = (processes, tiebreakerComparator) => {
  processes = processes
    .map(process => process.clone())
    .sort(createArrivalTimeComparator(tiebreakerComparator));

  let burstTimeComparator = createBurstTimeComparator(tiebreakerComparator);
  let totalExecutionTime = processes.reduce(
    (total, processOne) => total + processOne.burstTime,
    0
  );

  for (
    let currentExecutionTime = 0;
    currentExecutionTime < totalExecutionTime;
    currentExecutionTime++
  ) {
    let process = getProcessToExecute(
      processes,
      currentExecutionTime,
      burstTimeComparator
    );

    process.executions = process.executions || [];
    if (process.executions.length > 0) {
      let lastExecution = process.executions[process.executions.length - 1];
      if (lastExecution.end === currentExecutionTime) {
        lastExecution.end++;
      } else {
        process.executions.push({
          start: currentExecutionTime,
          end: currentExecutionTime + 1
        });
      }
    } else {
      process.executions.push({
        start: currentExecutionTime,
        end: currentExecutionTime + 1
      });
    }
    process.burstTime--;
  }

  let totalWaitingTime = 0;
  processes.forEach(process => {
    for (
      let executionIndex = 0;
      executionIndex < process.executions.length;
      executionIndex++
    ) {
      let execution = process.executions[executionIndex];
      if (executionIndex === 0) {
        totalWaitingTime += execution.start - process.arrivalTime;
      } else {
        let previousExecution = process.executions[executionIndex - 1];
        totalWaitingTime += execution.start - previousExecution.end;
      }
    }
  });

  return totalWaitingTime / processes.length;
};

module.exports = srtf;

File: java/src/br/com/brunohpaiva/osprocessscheduling/algorithms/Srtf.java
package br.com.brunohpaiva.osprocessscheduling.algorithms;

import br.com.brunohpaiva.osprocessscheduling.comparators.ArrivalTimeComparator;
import br.com.brunohpaiva.osprocessscheduling.comparators.BurstTimeComparator;
import br.com.brunohpaiva.osprocessscheduling.models.Process;

import java.util.Comparator;
import java.util.List;

public class Srtf {

    public static Double calculate(List<Process> processes, Comparator<Process> tiebreakerComparator) {
        processes.sort(new ArrivalTimeComparator(tiebreakerComparator));

        Comparator<Process> burstTimeComparator = new BurstTimeComparator(tiebreakerComparator);
        int totalExecutionTime = getTotalExecutionTime(processes);

        for (int currentExecutionTime = 0; currentExecutionTime < totalExecutionTime; currentExecutionTime++) {
            Process process = getProcessToExecute(processes, currentExecutionTime, burstTimeComparator);

            List<Process.Execution> executions = process.getExecutions();
            Process.Execution lastExecution = !executions.isEmpty() ? executions.get(executions.size() - 1) : null;
            if (lastExecution != null && lastExecution.getEnd() == currentExecutionTime)
                lastExecution.setEnd(lastExecution.getEnd() + 1);
            else
                executions.add(new Process.Execution(currentExecutionTime, currentExecutionTime + 1));

            process.setBurstTime(process.getBurstTime() - 1);
        }

        return getTotalWaitingTime(processes) / processes.size();
    }

    public static Process getProcessToExecute(List<Process> processes, int currentExecutionTime, Comparator<Process> burstTimeComparator) {
        return processes.stream().filter(process -> process.getArrivalTime() <= currentExecutionTime && process.getBurstTime() != 0).min(burstTimeComparator).orElse(null);
    }

    public static int getTotalExecutionTime(List<Process> processes) {
        return processes.stream().mapToInt(Process::getBurstTime).sum();
    }

    public static double getTotalWaitingTime(List<Process> processes) {
        return processes.stream().mapToDouble(process -> {
            int sum = 0;
            for (int index = 0; index < process.getExecutions().size(); index++)
                sum += process.getExecutions().get(index).getStart() -
                        (index == 0 ? process.getArrivalTime() : process.getExecutions().get(index - 1).getEnd());
            return sum;
        }).sum();
    }

}
